lib/elixir_sense/core/compiler.ex

```diff
@@ -0,0 +1,4696 @@
+    %{vars: {read, _write}, prematch: prematch} = s
+    pair = {name, var_context(meta, kind)}
+
+    result =
+      case read do
+        %{^pair => current_version} ->
+          case prematch do
+            {pre, _counter, {_bitsize, original}} ->
+              cond do
+                Map.get(pre, pair) != current_version ->
+                  {:ok, current_version}
+
+                Map.has_key?(pre, pair) ->
+                  # elixir plans to remove this case on 2.0
+                  {:ok, current_version}
+
+                not Map.has_key?(original, pair) ->
+                  {:ok, current_version}
+
+                true ->
+                  :raise
+              end
+
+            _ ->
+              {:ok, current_version}
+          end
+
+        _ ->
+          prematch
+      end
+
+    case result do
+      {:ok, pair_version} ->
+        var = {name, [{:version, pair_version} | meta], kind}
+        s = add_var_read(s, var)
+        {var, s, e}
+
+      error ->
+        case Keyword.fetch(meta, :if_undefined) do
+          {:ok, :apply} ->
+            # convert to local call
+            expand({name, meta, []}, s, e)
+
+          # elixir plans to remove this clause on v2.0
+          {:ok, :raise} ->
+            # elixir raises here undefined_var
+            {{name, meta, kind}, s, e}
+
+          # elixir plans to remove this clause on v2.0
+          _ when error == :warn ->
+            # convert to local call and add if_undefined meta
+            expand({name, [{:if_undefined, :warn} | meta], []}, s, e)
+
+          _ when error == :pin ->
+            # elixir raises here undefined_var_pin
+            {{name, meta, kind}, s, e}
+
+          _ ->
+            # elixir raises here undefined_var and attaches span meta
+            {{name, meta, kind}, s, e}
+        end
+    end
+  end
+
+  # Local calls
+
+  defp do_expand({fun, meta, args}, state, env)
+       when is_atom(fun) and is_list(meta) and is_list(args) do
+    # elixir checks here id fall is not ambiguous
+    arity = length(args)
+
+    # If we are inside a function, we support reading from locals.
+    allow_locals = match?({n, a} when fun != n or arity != a, env.function)
+
+    case NormalizedMacroEnv.expand_import(env, meta, fun, arity,
+           trace: false,
+           allow_locals: allow_locals,
+           check_deprecations: false
+         ) do
+      {:macro, module, callback} ->
+        # NOTE there is a subtle difference - callback will call expander with state derived from env via
+        # :elixir_env.env_to_ex(env) possibly losing some details. Jose Valim is convinced this is not a problem
+        state =
+          state
+          |> add_call_to_line({module, fun, length(args)}, meta)
+          |> add_current_env_to_line(meta, env)
+
+        expand_macro(meta, module, fun, args, callback, state, env)
+
+      {:function, module, fun} ->
+        {ar, af} =
+          case __MODULE__.Rewrite.inline(module, fun, arity) do
```

Comment:
Do you need to inline as well? It is an optimization, you should not need to care about it.

Reply:
I think I do. Otherwise I need to deal with both inlined and not inlined versions as `Kernel` macros return inlined versions when expanded

Reply:
The question though is: do you care about the result of the expansion? My understanding is that you'd only care about how it affects the environment but you don't need the actual expansion result, which should help reduce some of the complexity here.